pilot/pkg/model/push_context.go

```diff
@@ -259,18 +261,33 @@ func (first *PushRequest) Merge(other *PushRequest) *PushRequest {
 		Reason: append(first.Reason, other.Reason...),
 	}
 
-	// Only merge EdsUpdates when incremental eds push needed.
-	if !merged.Full {
-		merged.EdsUpdates = make(map[string]struct{})
-		// Merge the updates
-		for update := range first.EdsUpdates {
-			merged.EdsUpdates[update] = struct{}{}
+	// Do not merge when any one is nil
+	if first.ConfigsUpdated != nil && other.ConfigsUpdated != nil {
+		merged.ConfigsUpdated = make(map[resource.GroupVersionKind]map[string]struct{})
+		for update := range first.ConfigsUpdated {
+			merged.ConfigsUpdated[update] = nil
 		}
-		for update := range other.EdsUpdates {
-			merged.EdsUpdates[update] = struct{}{}
+		for update := range other.ConfigsUpdated {
+			merged.ConfigsUpdated[update] = nil
+		}
+
+		for kind := range merged.ConfigsUpdated {
+			m1, f1 := first.ConfigsUpdated[kind]
+			m2, f2 := other.ConfigsUpdated[kind]
+			// If kind presents in both, do not merge when any one is nil or empty because values
+			// has special meanings(for example, full push), else only merge non nil config maps
+			if (f1 && f2 && (len(m1) > 0 && len(m2) > 0)) || (!f1 || !f2 && (m1 != nil || m2 != nil)) {
```

Comment:
It is not a good idea to distinguish nil and length 0 in go. It makes developers tought to understand

Reply:
Yes, I have no clue what this line is doing, its very complex

Reply:
The logic looks complex because we use "nil" to represent special action(like full push). Today, the merging logic is that complex now...I'll try to do refactoring. I strongly do not recommend using "nil" to represent any business logic any more.